packages/discord.js/src/structures/Poll.js

```diff
@@ -0,0 +1,106 @@
+'use strict';
+
+const { Collection } = require('@discordjs/collection');
+const { Routes } = require('discord-api-types/v10');
+const Base = require('./Base');
+const { PollAnswer } = require('./PollAnswer');
+const { DiscordjsError } = require('../errors/DJSError');
+const { ErrorCodes } = require('../errors/index');
+
+/**
+ * Represents a Poll
+ * @extends {Base}
+ */
+class Poll extends Base {
+  constructor(client, data, message) {
+    super(client);
+
+    /**
+     * The message that started this poll
+     * @name Poll#message
+     * @type {Message}
+     * @readonly
+     */
+
+    Object.defineProperty(this, 'message', { value: message });
+
+    /**
+     * The question text of this poll
+     * @type {string}
+     */
+    this.question = data.question.text;
+
+    /**
+     * The answers of this poll
+     * @type {Collection<number, PollAnswer>}
+     */
+    this.answers = data.answers.reduce(
+      (acc, answer) => acc.set(answer.answer_id, new PollAnswer(this.client, answer, this)),
+      new Collection(),
+    );
+
+    /**
+     * The timestamp when this poll expires
+     * @type {number}
+     */
+    this.expiresTimestamp = Date.parse(data.expiry);
+
+    /**
+     * Whether this poll allows multiple answers
+     * @type {boolean}
+     */
+    this.allowMultiselect = data.allow_multiselect;
+
+    /**
+     * The layout type of this poll
+     * @type {PollLayoutType}
+     */
+    this.layoutType = data.layout_type;
+
+    this._patch(data);
+  }
+
+  _patch(data) {
+    if (data.results) {
+      /**
+       * Whether this poll's results have been precisely counted
+       * @type {boolean}
+       */
+      this.resultsFinalized = data.results.is_finalized;
+
+      for (const answerResult of data.results.answer_counts) {
+        const answer = this.answers.get(answerResult.id);
+        answer._patch(answerResult);
+      }
+    } else {
+      this.resultsFinalized ??= false;
+    }
+  }
+
+  /**
+   * The date when this poll expires
+   * @type {Date}
+   * @readonly
+   */
+  get expiresAt() {
+    return new Date(this.expiresTimestamp);
+  }
+
+  /**
+   * End this poll
+   * @returns {Promise<Message>}
+   */
+  async end() {
+    if (Date.now() > this.expiresTimestamp) {
+      throw new DiscordjsError(ErrorCodes.PollAlreadyExpired);
+    }
+
+    const message = await this.client.rest.post(Routes.expirePoll(this.message.channel.id, this.message.id));
+
+    this.message._patch(message);
+
+    return this.message;
```

Comment:
Won't this cause the GW event to emit a wrong messageUpdate? 👀 
We might need to clone this.message and patch the clone instead?